matchtools/_matchblock.py

```diff
@@ -11,91 +11,187 @@
 from geopy.distance import great_circle
 from geopy.distance import vincenty
 
+__all__ = ['MatchBlock']
 
-def check_tolerance(*dec_args):
-    """Wrapper checking whether tolerance is within the bounds."""
-    def wrapper(f):
+
+def tolerance_interval(lower=0, upper=None):
+    """Decorator checking whether tolerance is within the boundaries."""
+
+    def decorator(f):
         @wraps(f)
-        def wrapped(*args, **kwargs):
+        def wrapper(*args, **kwargs):
             if 'tolerance' in kwargs:
-                if lower_bound is not None and upper_bound is None:
-                    if kwargs['tolerance'] < lower_bound:
+                if lower is not None and upper is None:
+                    if kwargs['tolerance'] < lower:
                         msg = 'tolerance must be higher than {}'
-                        raise ValueError(msg.format(lower_bound))
-                elif lower_bound is None and upper_bound is not None:
-                    if kwargs['tolerance'] > upper_bound:
+                        raise ValueError(msg.format(lower))
+
+                elif lower is None and upper is not None:
+                    if kwargs['tolerance'] > upper:
                         msg = 'tolerance must be lower than {}'
-                        raise ValueError(msg.format(upper_bound))
-                elif lower_bound is not None and upper_bound is not None:
-                    if not (lower_bound <= kwargs['tolerance'] <= upper_bound):
+                        raise ValueError(msg.format(upper))
+
+                elif lower is not None and upper is not None:
+                    if not (lower <= kwargs['tolerance'] <= upper):
                         msg = 'tolerance must be in between {}, and {}'
-                        raise ValueError(msg.format(lower_bound, upper_bound))
+                        raise ValueError(msg.format(lower, upper))
             return f(*args, **kwargs)
-        return wrapped
+        return wrapper
+
+    if callable(lower):
+        func, lower = lower, 0
+        return decorator(func)
+    return decorator
+
 
-    if len(dec_args) == 1 and callable(dec_args[0]):
-        # no bounds provided
-        lower_bound, upper_bound = 0, None
-        return wrapper(dec_args[0])
+class Tolerance:
+    """Tolerance attributes descriptor."""
 
-    if len(dec_args) == 2 and callable(dec_args[1]):
-        # only lower bound provided
-        lower_bound, upper_bound = dec_args[0], None
-    elif len(dec_args) == 2:
-        # both bounds provided
-        lower_bound, upper_bound = dec_args[0], dec_args[1]
-    else:
-        raise ValueError('too many arguments: {}'.format(dec_args))
+    def __init__(self, attribute):
+        self.attribute = '_' + attribute
```

Comment:
If your only concern regarding this is speed, I think it's not an issue here: this operation performed only once, and we are not joining tens or hundreds of strings together, where join() is the preferred way. In our case though, I think readability is better.

In fact, I have even made a quick measurement here, and the result is quite surprising to me:

```
In [1]: %timeit ''.join(('_', 'attribute'))
159 ns ± 3.33 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [2]: %timeit '_' + 'attribute'
17.2 ns ± 0.0439 ns per loop (mean ± std. dev. of 7 runs, 100000000 loops each)
```

So it turns out, that in current CPython implementation, it's actually faster to join two strings together using addition operator rather that string join() method.

Also, I like this short [answer](https://softwareengineering.stackexchange.com/a/304621) regarding string concatenation problem.